src/session/session-utils.c

```diff
@@ -638,3 +638,64 @@ spawn_and_wait (const char **argv, const char **envp,
       return wstatus;
     }
 }
+
+
+char *
+run_user_command (const char *cmd, const char *rhost, const char *authorization)
```

Comment:
Is it *really* necessary to introduce a new function for that? There are a lot of things wrong with it, and we can go through it one by one -- but it seems to me that you should be able to use the existing `spawn_and_wait()` for this? You can remap fd 1 (stdout) to a `pipe()` and read the user name from that. Now, I don't know if you can actually do that *after* the process already exited (could work for small amounts of data depending on how much Linux buffers, but you only expect a user name). If not, then let's rather expand `spawn_and_wait()` to collect the output than having a similar duplicate function.